### Classification approach for exercise

In [3]:
import os,re
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
from sklearn import metrics
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_absolute_error,mean_squared_error
from sklearn.metrics import classification_report,confusion_matrix,log_loss
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier

pd.set_option('display.max_columns',200)
pd.set_option("display.max_rows",500)

In [5]:
dfTrain = pd.read_csv("exercise_40_train.csv")
dfTest = pd.read_csv("exercise_40_test.csv")
dfTest.head(15)


x1         x2         x3        x4        x5        x6        x7  \
0   4.747627  20.509439  Wednesday  2.299105 -1.815777 -0.752166   0.0098%   
1   1.148654  19.301465        Fri  1.862200 -0.773707 -1.461276   0.0076%   
2   4.986860  18.769675   Saturday  1.040845 -1.548690  2.632948   -5e-04%   
3   3.709183  18.374375    Tuesday -0.169882 -2.396549 -0.784673   -0.016%   
4   3.801616  20.205541     Monday  2.092652 -0.732784 -0.703101   0.0186%   
5   1.449185  17.237848        Sun  0.998924  0.882350 -1.700730   0.0076%   
6   0.979035  18.056038     Monday  0.105417 -1.782850 -1.390759  -0.0045%   
7   0.910023  17.057286        Mon -0.903745  0.118699 -2.110974  -0.0079%   
8   3.616287  17.740881       Thur  2.193121  4.222402  0.206030  -0.0201%   
9   2.539012  21.078729   Saturday -0.868620 -0.193794  1.244369  -0.0222%   
10  2.202450  20.431475        Mon  0.639342 -2.867947 -1.407299   -3e-04%   
11  6.334361  19.658003     Friday  0.555120 -0.096129  0.861986  -0.0066%   
12  4.047808  23.034961        Sat -0.505224 -0.864426  0.115058  -0.0158%   
13  3.078519  21.370217  Wednesday -0.918152 -0.777777  0.695998  -0.0096%   
14  4.768995  21.048787  Wednesday -0.183138 -0.959152  0.501082  -0.0195%   

          x8        x9       x10         x11        x12       x13       x14  \
0  -3.240309  0.587948 -0.260721  101.113628  -0.812035  3.251085 -0.004432   
1   0.443209  0.522113 -1.090886  104.791999   8.805876  1.651993       NaN   
2  -1.167885  5.739275  0.222975  102.109546   7.831517  3.055358  2.036434   
3  -2.662226  1.548050  0.210141   82.653354   0.436885  1.578106       NaN   
4   0.056422  2.878167 -0.457618   75.036421   8.034303  1.631426  0.643738   
5  -0.371594  1.269817 -1.433127   78.434085   3.124643  4.438999       NaN   
6  -1.886515  2.905928  0.222237   87.040247   6.739816  2.839826  0.227705   
7  -0.613028  2.109169  1.140936   93.985642   1.006534  2.645002  0.938672   
8  -0.397257  3.058279 -0.859673  120.832847   8.501188  2.386375  0.841161   
9  -0.334661  1.478947  0.637950   88.678380  -1.291054  2.964086  0.268776   
10  1.538257  3.544400 -0.965467  101.354472   8.760391  5.464629  0.733497   
11  0.594773  1.582971  1.478990   97.004902  -2.995886  3.045324  0.115462   
12 -2.283415  1.985695 -1.769429  105.941368  10.299181  2.836541  0.835711   
13  0.632970  2.430308  1.253636  105.366919   3.839965  6.779266  0.994631   
14  2.752925  5.526124 -1.602605  122.870329   7.220539  2.991452       NaN   

         x15         x16        x17        x18                 x19       x20  \
0  -0.917763  102.193597   7.097387  75.256714   $120.216189955777  8.269754   
1  -1.396065   99.356609   7.117930  79.061540  $-267.562586413086  5.668347   
2   1.057296         NaN  10.943217  87.567827  $-311.292903116571  3.219583   
3  -1.287913  102.410965   6.588790  71.825782   $2229.14940030076  7.459929   
4   0.349166  101.513490   5.777599  74.602441  $-469.049529991235  8.245194   
5  -0.761186  102.214681  10.767388  77.358212   $1052.65320492622  4.483987   
6  -3.909978  100.183345   6.302176  81.401764  $-1248.56374994913  7.149937   
7  -1.301097  100.798786   5.676887  86.843134   $2192.44913918191  7.323247   
8   2.918181   99.884976   3.951736  71.032462   $-744.50837755944  8.676654   
9  -1.237706         NaN   7.867055  89.714961   $621.500064669709  5.846521   
10  0.618157         NaN  10.094763  90.123582   $1641.48095109299  0.695014   
11 -0.080587         NaN   7.524240  80.905777  $-2803.98880622137  9.902445   
12  1.468347   99.542627   4.883594  73.840790  $-960.320565434093 -0.011208   
13  0.899184         NaN  10.035408  81.874254  $-360.765613820984  5.393525   
14 -0.610510  101.378570   9.087118  78.425467  $-743.622584295803  2.311300   

          x21         x22       x23     x24        x25       x26          x27  \
0    4.794438   56.134458 -0.083641     NaN   5.740955 -3.152540   924.013304   
1   12.377178         NaN  0.321981  female   2.05812

### Combine data frames

#### Goal is to encode categorical variables for train & test

In [7]:
dfCombine = pd.concat([dfTrain,dfTest])
dfCombine.iloc[40000:40001,]

y        x1         x2         x3        x4        x5        x6       x7  \
0 NaN  4.747627  20.509439  Wednesday  2.299105 -1.815777 -0.752166  0.0098%   

         x8        x9       x10         x11       x12       x13       x14  \
0 -3.240309  0.587948 -0.260721  101.113628 -0.812035  3.251085 -0.004432   

        x15         x16       x17        x18                x19       x20  \
0 -0.917763  102.193597  7.097387  75.256714  $120.216189955777  8.269754   

        x21        x22       x23  x24       x25      x26         x27  \
0  4.794438  56.134458 -0.083641  NaN  5.740955 -3.15254  924.013304   

         x28       x29  x30 x31       x32      x33        x34     x35  \
0  17.697125  1.402273  NaN  no  1.461654  Florida  57.930285  4.7273   

         x36        x37       x38         x39        x40           x41  \
0  66.416594  28.450466  33.49931  5-10 miles  16.776061  99971.844703   

        x42       x43  x44  x45       x46        x47        x48  x49  \
0  0.492812 -0.963616  NaN  NaN  1.266416  53.020326  61.840284  NaN   

           x50       x51  x52        x53  x54       x55       x56  x57  \
0  1825.413159  2.517924  NaN  70.880778  NaN  1.923125  0.239009  NaN   

         x58  x59  x60       x61       x62       x63        x64          x65  \
0  300.62999    0  May  5.461123  5.149473  1.179229  59.346822  progressive   

        x66        x67       x68       x69       x70       x71        x72  \
0  2.219502  17.667792  5.557066 -2.030901  0.636111 -0.000234  62.363381   

        x73       x74       x75       x76       x77        x78  x79       x80  \
0  4.613878  2.415655  3.632041 -0.382482  mercedes  55.938387  1.0  8.325299   

        x81       x82      x83      x84  x85      x86       x87       x88  \
0 -1.274085 -5.663245  2.40266 -0.06104  NaN  64.0025  0.548974  3.420875   

         x89       x90  x91        x92 x93       x94       x95        x96  \
0  11.553023  5.904644  NaN  12.542333  no  3.107683  0.533904  12.438759   

        x97  x98  x99      x100  
0  7.298306    0  NaN  93.56712

In [8]:
dfCombine.reset_index(drop=True,inplace=True)
dfCombine.tail(1)

y        x1         x2      x3        x4        x5        x6       x7  \
49999 NaN  2.972268  20.485504  Sunday  1.747878  0.197357  1.798676  0.0086%   

             x8        x9      x10        x11       x12       x13       x14  \
49999 -0.318751  2.858806  0.08416  99.642148  8.497151  0.878065 -0.338883   

            x15         x16       x17       x18                 x19       x20  \
49999  0.350812  101.065927  7.900097  76.79105  $-508.049552925335  0.443446   

            x21        x22       x23     x24       x25  x26         x27  \
49999 -0.255657  82.889762 -2.112773  female  2.764069  NaN  1028.57299   

            x28       x29       x30  x31       x32  x33        x34       x35  \
49999  18.45714  2.473741 -0.578496  yes  0.634841  NaN  50.916587  1.271283   

             x36        x37        x38         x39        x40           x41  \
49999  63.837508  29.870539  39.909804  5-10 miles  19.938424  101720.09239   

            x42       x43  x44       x45       x46        x47        x48  \
49999 -0.875818  0.765511  NaN  0.996541 -3.489898  49.483851  65.862504   

            x49         x50       x51       x52        x53      x54       x55  \
49999  7.303529  1150.95107  1.808276  9.110502  72.228876 -0.61096  2.634673   

           x56  x57        x58  x59        x60       x61       x62       x63  \
49999 -1.92103  NaN  300.62999    0  September  6.847741  2.181149  1.297841   

             x64          x65       x66        x67        x68      x69  \
49999  50.885163  progressive  0.153248  14.402071  11.865657 -1.29376   

            x70       x71        x72       x73  x74       x75       x76  x77  \
49999  0.434589 -0.000234  52.004365  4.865971  NaN  1.524051  0.731896  NaN   

             x78  x79       x80       x81       x82       x83       x84  \
49999  45.800511  1.0  7.780886  0.953972 -6.183223  1.565181  3.140183   

             x85  x86       x87       x88        x89       x90       x91  \
49999  95.665994  NaN -0.127836 -1.550846  11.222902  4.651827  1.397517   

            x92  x93       x94       x95  x96        x97  x98  x99       x100  
49999  8.463199  yes  3.942066  1.245289  NaN  11.718582    0  yes  98.307681

### Clean data

#### Removal of null values

In [9]:
dfCombine.isnull().sum()

y       10000
x1          0
x2          0
x3          0
x4          0
x5       3030
x6          0
x7          0
x8          0
x9          0
x10         0
x11      6439
x12         0
x13         0
x14     12292
x15         0
x16     13965
x17         0
x18         0
x19         0
x20         0
x21         0
x22      3000
x23         0
x24      4825
x25         0
x26      3050
x27         0
x28         0
x29         0
x30     40421
x31         0
x32         0
x33      8941
x34         0
x35         0
x36         0
x37         0
x38      2967
x39         0
x40         0
x41     11907
x42     12147
x43         0
x44     42813
x45     10072
x46         0
x47         0
x48         0
x49     16077
x50         0
x51         0
x52     20262
x53         0
x54     15934
x55     22120
x56         0
x57     40541
x58         0
x59         0
x60         0
x61      9059
x62         0
x63      3007
x64      6363
x65         0
x66         0
x67      3045
x68      2984
x69         0
x70         0
x71   

### Drop columns with null values & other un-needed columns

In [10]:
# dfCombine.drop(["x5","x11","x14","x16","x22","x24","x26","x30","x33","x38","x41","x42","x44","x45","x49","x52","x54",\
# "x55","x57","x61","x63","x64","x67","x68","x74","x75","x76","x77","x78","x79","x80","x85","x89","x95","x99",\
# "x83","x86","x88","x91","x92","x94","x96"],axis=1,inplace=True)
# unneeded: x39
dfCombine.drop(["x11","x14","x16","x24","x30","x33","x38","x39","x41","x42","x44","x45","x49","x52","x54",\
"x55","x57","x61","x64","x74","x75","x76","x77","x78","x79","x80","x85","x89","x95","x99",\
"x83","x86","x88","x91","x96"],axis=1,inplace=True)

In [11]:
dfCombine.isnull().sum()

y       10000
x1          0
x2          0
x3          0
x4          0
x5       3030
x6          0
x7          0
x8          0
x9          0
x10         0
x12         0
x13         0
x15         0
x17         0
x18         0
x19         0
x20         0
x21         0
x22      3000
x23         0
x25         0
x26      3050
x27         0
x28         0
x29         0
x31         0
x32         0
x34         0
x35         0
x36         0
x37         0
x40         0
x43         0
x46         0
x47         0
x48         0
x50         0
x51         0
x53         0
x56         0
x58         0
x59         0
x60         0
x62         0
x63      3007
x65         0
x66         0
x67      3045
x68      2984
x69         0
x70         0
x71         0
x72         0
x73         0
x81         0
x82         0
x84         0
x87         0
x90         0
x92      3061
x93         0
x94      2955
x97         0
x98         0
x100        0
dtype: int64

In [13]:
dfCombine[["x5","x22","x26","x63","x67","x68","x92","x94"]].sample(30)

x5         x22       x26       x63        x67        x68  \
21763       NaN   19.470680       NaN -1.274969  14.402071  10.029268   
27925 -1.034614   86.954794 -1.207133  0.175002  14.402071        NaN   
12731  1.353793  155.434208  2.040047  0.399203  14.402071   5.258130   
34984  0.874903   29.957684 -0.177468 -0.892646        NaN  19.819020   
5811  -0.141493   30.563129  0.390292  0.105798  14.402071  21.162576   
29576  1.324089   43.528315  3.219540  0.876129  14.402071  23.717844   
34851  2.016772   59.978897  0.535555  0.569671  14.402071  21.025405   
11780 -0.000970   71.073110  2.575249 -1.148309  17.504057  19.307316   
22476  2.028108   56.475893  3.107794 -2.686838   6.849080  16.258699   
25358  1.627839  177.917257  1.991796 -1.131329  14.402071  17.530436   
45021       NaN   68.079973  0.263596 -1.885047  14.402071  15.100455   
19238  0.098876   47.495568       NaN  0.504898        NaN        NaN   
938    1.882768   67.430360  0.140454  0.409262  14.402071  17.972258   
42407 -0.023689   65.099149  0.308102  0.681704  14.402071  14.394431   
6850  -1.934099   70.754539       NaN -0.982058  17.205226  21.692060   
26914       NaN   73.094073  1.675443 -1.899023  14.402071   7.498441   
4314   2.357386   63.262110  2.100539  1.383504  14.402071  11.131099   
1032        NaN   54.711343 -0.273298  0.842289        NaN  17.090878   
4287  -0.264132   78.119287 -2.283954  2.778058  14.402071  25.741297   
26000 -1.668831  128.785525 -3.310046  2.315139  14.402071  20.751448   
1489   1.663074         NaN  1.419542 -1.928856  14.402071  21.341839   
19989 -1.528951   67.262947  2.226410 -0.953478  14.402071   6.910976   
33736       NaN   18.782722 -2.550753 -1.044709  14.402071   9.077822   
7004        NaN   52.795246 -0.925301 -1.002171  14.402071  22.231309   
6806        NaN   26.545551  1.295858 -0.332447  14.402071  18.594385   
37342 -0.173910  122.356103 -0.901542 -1.949076  14.402071  21.551679   
40100 -0.118121  133.347191 -0.105452 -1.379292  12.781655  18.948623   
49594 -0.687420   64.081215       NaN       NaN  12.711906   8.749202   
13694  1.732185   57.215464  0.250021 -2.353327  14.402071  11.196956   
39355 -0.899189   50.618307  0.246583 -4.446399  14.402071  15.375114   

            x92       x94  
21763  1.852088  0.633346  
27925  8.395535  3.295258  
12731  4.369000  3.359061  
34984  3.587513  2.138184  
5811   3.043976  2.662600  
29576  5.336864  4.719040  
34851  2.090957  2.329912  
11780       NaN  2.431110  
22476  2.827361  5.551139  
25358  2.551657       NaN  
45021  4.770238  3.712546  
19238  7.251258  2.503586  
938    6.575382  3.610960  
42407       NaN  5.765002  
6850   6.225296  1.285016  
26914  3.520931  3.266654  
4314  -0.927437  3.065385  
1032   5.330014  3.397137  
4287   6.933513  1.506992  
26000  3.962585  5.335641  
1489   5.971544  2.532597  
19989  4.028269  1.687927  
33736  2.717216       NaN  
7004   4.393595  5.942950  
6806   6.317518  5.192863  
37342  0.358605  2.727518  
40100  3.031339  4.697883  
49594  4.000392  3.394416  
13694  4.235439       NaN  
39355  2.378059  2.438867

### Interpolate missing values

In [14]:
dfCombine.interpolate(method="linear",inplace=True)
dfCombine.sample(25)

y        x1         x2         x3        x4        x5        x6  \
24802  1.0  3.700801  19.555568     Monday -1.532088 -0.923755  0.111407   
43069  0.0  1.649476  19.615851    Tuesday -1.115213  2.141787 -1.129252   
39750  0.0  4.439545  18.472721   Saturday  0.363669 -0.169525 -0.616954   
17480  0.0  4.599109  19.742917       Thur -0.557934 -0.329235  0.970999   
7325   0.0  1.735975  20.905010        Sat -2.070306  0.627281  0.974494   
28234  1.0  2.135756  18.079309     Monday  0.193932 -0.518779 -0.845606   
10906  0.0  2.314360  20.653557     Monday -1.463253 -0.170908 -0.940822   
26155  0.0  1.727891  19.953020        Tue  3.059611 -1.270675  0.503054   
37106  0.0  1.987373  16.892603   Saturday  0.062117  0.831228 -3.573213   
16850  1.0  3.811771  19.145806   Saturday  0.544372  1.288862 -1.796131   
36685  1.0  5.503534  21.277037  Wednesday -0.717922  0.463077  2.229445   
36919  0.0  3.142634  19.478533   Thursday  0.246668 -0.206105 -0.235294   
42052  0.0  5.422859  18.390780     Sunday -0.547900 -2.261003 -2.236781   
48036  0.0  4.538081  20.110792        Sat  3.368798  0.075786 -1.934761   
21630  0.0  6.545719  20.175982   Saturday -3.263711  0.235463  0.956779   
36017  0.0  3.258399  16.876443        Fri -0.590817  0.750829 -0.106237   
11999  0.0  1.702191  19.568247        Tue -0.441816 -3.037061  0.412381   
4261   0.0  3.169784  20.138623        Sat  1.281018  2.289340 -1.027467   
22807  1.0  7.986880  19.786592       Thur  0.346689 -1.052555  1.267785   
22365  0.0 -0.542396  22.378060        Tue -0.991832  0.307687 -0.034229   
4414   0.0  6.172124  19.385102     Sunday  0.028463 -1.094615 -0.278041   
23986  0.0  2.874729  17.685294    Tuesday  3.861115 -0.845363  1.458322   
23042  0.0  2.314455  20.228447     Friday -3.872280 -1.869012 -2.689611   
3399   0.0  3.048022  22.907371    Tuesday -0.658377 -0.808474 -0.611657   
21699  0.0  2.701669  19.675842   Saturday -0.643488 -1.471979 -1.056891   

             x7        x8        x9       x10       x12       x13       x15  \
24802   -0.001% -1.182973  8.051184  1.041465  7.703282  7.457535  1.340633   
43069   -0.006%  0.569842  2.571683  1.035530  6.862350  1.606349 -0.107158   
39750    0.014% -1.528909  3.701774 -0.136310  8.338010  3.874768  0.517432   
17480   0.0101%  0.271654  6.380765 -0.191324  1.844777  4.559143  0.608429   
7325    0.0097% -0.320903  5.200087 -0.233427 -0.923607  5.638686  0.936419   
28234   -0.008% -1.164756  0.179716  0.893962 -0.232957  5.375917 -0.965631   
10906   0.0037%  1.911146  2.400728 -0.739005  3.799590  3.238841  0.635056   
26155   0.0058%  3.214256  3.319155  0.496622 -2.340959  1.428795  1.689848   
37106   0.0155% -3.274020  5.364292 -1.437268  9.627622  4.009394 -0.147932   
16850   0.0041%  0.692481  2.053905  0.829416  7.789863  6.698539  0.489762   
36685  -0.0177% -0.868476  1.761435 -1.227574  7.809941  6.521363 -0.173944   
36919   0.0155% -0.907021  2.456190  0.329134 -1.636812  3.624262  2.500112   
42052  -0.0014% -0.914101  4.606962  1.052603  7.602467  5.460721 -0.697242   
48036  -0.0045%  0.275940  1.652981  1.254639  2.874947  3.633881  1.443148   
21630    6e-04%  1.222570  5.657776  0.563531  7.148409  6.753268  0.660187   
36017   0.0214% -0.215632  5.307602  0.345362  5.042832  2.555252 -0.651464   
11999   0.0106%  1.617127  4.870122  1.336953  2.424745  2.011211 -0.275917   
4261    0.0192%  0.307978  2.198982  0.166519  0.703358  1.758176 -0.548451   
22807   -0.004%  1.231159  2.304579 -0.769884  2.129733  2.257558  0.127386   
22365  -0.0029% -2.248374  0.827209 -0.357681 -0.734158  2.942921 -1.657539   
4414    0.0061%  1.488316  0.344042 -0.932483  3.122444  2.949748 -0.284601   
23986   -0.012%  0.256347  2.912029  0.131703  6.509785  0.971462 -2.442526   
23042  -0.0103% -1.520306  3.724672  1.919978 -0.648051  3.658704  1.794834   
3399   -0.0051% -1.286155  0.299909  1.031937 -1.332368  0.221184  0.957981   
21699   0.0092% -0.545805  2.903073  0.545468  6.476878  

In [15]:
dfCombine.isnull().sum()

y       0
x1      0
x2      0
x3      0
x4      0
x5      0
x6      0
x7      0
x8      0
x9      0
x10     0
x12     0
x13     0
x15     0
x17     0
x18     0
x19     0
x20     0
x21     0
x22     0
x23     0
x25     0
x26     0
x27     0
x28     0
x29     0
x31     0
x32     0
x34     0
x35     0
x36     0
x37     0
x40     0
x43     0
x46     0
x47     0
x48     0
x50     0
x51     0
x53     0
x56     0
x58     0
x59     0
x60     0
x62     0
x63     0
x65     0
x66     0
x67     0
x68     0
x69     0
x70     0
x71     0
x72     0
x73     0
x81     0
x82     0
x84     0
x87     0
x90     0
x92     0
x93     0
x94     0
x97     0
x98     0
x100    0
dtype: int64

### x3
Clean up day names

In [16]:
dfCombine["x3"].unique()

array(['Wed', 'Friday', 'Thursday', 'Tuesday', 'Sunday', 'Saturday',
       'Sat', 'Wednesday', 'Sun', 'Tue', 'Thur', 'Monday', 'Fri', 'Mon'],
      dtype=object)

In [17]:
dfCombine["x3"] = dfCombine["x3"].replace(["Mon"],"Monday")
dfCombine["x3"] = dfCombine["x3"].replace(["Tue"],"Tuesday")
dfCombine["x3"] = dfCombine["x3"].replace(["Wed"],"Wednesday")
dfCombine["x3"] = dfCombine["x3"].replace(["Thur"],"Thursday")
dfCombine["x3"] = dfCombine["x3"].replace(["Fri"],"Friday")
dfCombine["x3"] = dfCombine["x3"].replace(["Sat"],"Saturday")
dfCombine["x3"] = dfCombine["x3"].replace(["Sun"],"Sunday")

In [18]:
dfCombine["x3"].unique()

array(['Wednesday', 'Friday', 'Thursday', 'Tuesday', 'Sunday', 'Saturday',
       'Monday'], dtype=object)

### x7
Remove percent sign

In [19]:
dfCombine['x7'] = dfCombine["x7"].str.replace("%","")

In [20]:
dfCombine.head(5)

y        x1         x2         x3        x4        x5        x6       x7  \
0  0.0  0.165254  18.060003  Wednesday  1.077380 -1.339233 -1.584341   0.0062   
1  1.0  2.441471  18.416307     Friday  1.482586  0.920817 -0.759931   0.0064   
2  1.0  4.427278  19.188092   Thursday  0.145652  0.366093  0.709962   -8e-04   
3  0.0  3.925235  19.901257    Tuesday  1.763602 -0.251926 -0.827461  -0.0057   
4  0.0  2.868802  22.202473     Sunday  3.405119  0.083162  1.381504   0.0109   

         x8        x9       x10       x12       x13       x15       x17  \
0  0.220784  1.816481  1.171788  4.644568  4.814885 -0.587361  4.278444   
1  1.192441  3.513950  1.419900  1.459868  1.443983 -1.165869  8.487132   
2  0.952323  0.782974 -1.247022  1.098525  1.216059  0.211685  8.601193   
3 -0.520756  1.825586  2.223038 -1.390239  3.962961 -2.046856  6.611554   
4 -0.732739  2.151990 -0.275406  7.230125  3.877312 -1.201565  9.103015   

         x18                 x19       x20        x21         x22       x23  \
0  77.709700  $-908.650758424405  7.328029  36.601967  126.384294  1.264713   
1  70.808566   $-1864.9622875143  6.796408   3.789176   93.037021  0.346201   
2  75.922820  $-543.187402955527  3.802395   7.407996   71.022413  0.070968   
3  74.966925  $-182.626380634258  7.728963   8.136213  121.610846  0.700954   
4  77.977156   $967.007090837503  2.272765  24.452102   56.492317 -1.102387   

        x25       x26          x27        x28       x29  x31       x32  \
0  3.834082  2.352406   905.491445  14.779467  0.752524   no -0.597288   
1  4.345028  1.651579   868.784447  25.914680  1.445294   no  1.127376   
2  4.332644 -0.375737  1077.233497  13.177479  4.174536   no  2.605279   
3  7.294990 -0.603983  1051.655489  17.006528  2.347355   no  1.071202   
4  6.313827  0.429187   949.904947  16.962710  0.510985  yes  2.283921   

         x34       x35        x36        x37        x40       x43       x46  \
0  46.667221  3.159002  69.444841  32.450076  18.129499  1.227266  0.234654   
1  47.928569 -1.013712  61.289132  35.521824  11.389081 -0.822055  1.104631   
2  56.674425  0.108486  69.798923  30.684074  14.372443  0.150728 -2.561083   
3  59.154933  1.319711  65.408246  34.401290  13.191173 -0.965711 -2.123048   
4  46.445617  0.022747  66.662910  31.135261  17.210430  1.012093  3.477451   

         x47        x48          x50       x51        x53       x56  \
0  53.359113  58.251700  1602.756464  2.740134  72.167186 -1.532562   
1  37.728834  56.052749  3350.837875  1.995635  73.964560  0.956001   
2  50.236892  63.975108  3756.910196  1.772648  59.428690  4.127857   
3  41.857197  59.226119  1961.609788  3.155214  68.671023  0.663759   
4  47.844153  55.921988  2345.195505  3.253079  71.853326 -2.273628   

          x58  x59        x60       x62       x63       x65       x66  \
0  300.629990    0     August  6.600008  1.252479   farmers  1.212048   
1  300.629990    0      April  3.864254  0.057936  allstate  0.738526   
2  182.369349    0  September  1.370661 -0.239705     geico -0.032980   
3  300.629990    0  September  0.013162  0.318335     geico -0.466535   
4  149.725023    0    January  1.305936  0.418272     geico -0.195764   

         x67        x68       x69       x70       x71        x72       x73  \
0  12.896733  15.263491  0.518653  0.543768 -0.912506  53.521544  7.723719   
1  14.402071  33.940951 -0.140384  1.016703 -0.000234  46.797571  7.260365   
2  14.402071  18.211817 -0.819798  1.010811 -0.000234  48.202036  9.336021   
3  14.402071  14.629914  1.389325  0.704880 -1.510949  49.882647  5.661421   
4  14.402071  12.227512 -2.951163  1.096548 -0.000234  51.349106  9.422401   

        x81       x82       x84       x87       x90       x92  x93       x94  \
0 -0.471594 -3.261865  0.383351 -0.162668  2.770844  0.909479   no  5.492487   
1  1.691428 -4.789183  2.138771  1.252022  3.703013  9.093466   no  3.346429   
2 -2.608974 -6.387984  2.138771  0.657292  3.997183  0.305170   no  4.456565   
3  1.424316 -5.43

In [21]:
# remove "E" notation
dfCombine["x7"] = pd.to_numeric(dfCombine["x7"])
dfCombine.sample(25)

y        x1         x2         x3        x4        x5        x6  \
772    0.0  0.107077  20.552642    Tuesday -1.437498  0.762980 -1.011265   
34307  0.0  1.273023  22.053456   Saturday -1.994792  1.087270 -0.567472   
45443  0.0  4.608894  20.290190    Tuesday  3.078879  0.150438 -1.433616   
44084  0.0  3.293018  19.157420   Saturday  1.420474 -0.219849  1.033762   
28490  0.0  3.217393  17.342641     Monday -1.478107 -1.953954  1.796983   
13044  1.0  2.235121  21.351481     Friday -0.408905  0.034197  1.207943   
11566  0.0  3.566892  18.354018    Tuesday  0.370415 -1.352613  1.786478   
6020   0.0  6.417232  16.267216  Wednesday  0.674452  0.517701  0.493084   
44518  0.0  5.598452  22.614382   Thursday  1.944444 -1.019805 -0.604954   
34632  0.0  3.204272  18.472658    Tuesday  1.959901 -0.984151 -1.275705   
36006  0.0  7.604115  22.679261  Wednesday  1.890199  0.857926  0.709843   
36692  0.0  1.170218  21.415162    Tuesday  2.685896 -0.515156 -0.061542   
37179  0.0  5.574849  20.648591    Tuesday -1.899940 -1.605405  1.113217   
13517  0.0  0.293778  21.420342     Sunday  1.452321  1.092927 -0.866923   
27838  0.0  3.869640  22.469852    Tuesday -0.886191  1.428220  0.609072   
37429  0.0  4.139252  17.848671     Sunday -1.995555  1.285428  0.491748   
45644  0.0  4.053205  19.283539   Saturday -1.556214  0.393511  0.057016   
30258  0.0  1.657434  23.638888  Wednesday  1.551072 -2.105190  1.713141   
17206  0.0  2.568067  21.362281    Tuesday  1.687603  1.156135  0.116361   
22095  1.0  2.085712  19.411830  Wednesday -1.204179 -1.968057  1.364603   
3386   0.0  1.326804  15.238900    Tuesday -2.370436  1.305366 -0.573418   
12353  0.0 -0.432037  22.669931   Thursday  1.637744 -1.687230  1.977945   
3432   0.0  6.404034  18.491963   Saturday -1.078288 -0.845044 -0.265682   
28592  0.0  2.723295  17.378849   Saturday  1.165603  1.284336 -0.215406   
1828   0.0  3.507543  22.705371  Wednesday -1.938616  0.214631 -1.895747   

           x7        x8        x9       x10       x12       x13       x15  \
772   -0.0041  1.070565  0.928667  2.869993  5.389915  2.871712  0.358573   
34307 -0.0217 -0.786781  1.053595  1.541803  4.117793  3.773072  1.463786   
45443 -0.0167 -1.871372  1.518364  0.424728  3.317760  1.860178 -1.093891   
44084  0.0029  1.672068  1.399318  0.592411  0.526290  6.739052 -2.400671   
28490  0.0142 -0.439407  5.409125  1.306892 -0.666420  4.782358  3.103740   
13044 -0.0043  0.006180  3.896382  0.818554  1.317243  1.281014 -0.115567   
11566  0.0107  0.497107  2.393512 -0.015477 -0.381765  2.673937 -0.163079   
6020   0.0014  1.117048  3.316549 -0.548476 -0.719201  1.576838 -0.677667   
44518  0.0040  1.528982  1.737960  0.585411  1.645548  0.827693 -0.204666   
34632  0.0146  0.753399 -0.347284  0.669267  5.975820  4.990730  1.284780   
36006  0.0100  0.613882  0.751073  0.006844  8.000049  7.416332 -1.553203   
36692  0.0142  0.328407  0.898711  0.042031  4.248853  2.795317  3.023937   
37179 -0.0142 -0.106678  2.781045  1.280861  5.441887  6.811431  0.622484   
13517 -0.0061 -2.288282  3.135281  1.242432  2.781183  4.518743  0.521805   
27838 -0.0059  1.316280  0.366222  1.518122  1.758677  5.399852  0.403615   
37429 -0.0076 -0.811499  3.829737 -1.240594 -0.282551  1.109278  0.930713   
45644  0.0011  0.634301  2.018157  1.028872 -2.129289  4.522182 -0.798533   
30258 -0.0113 -2.084626  2.128032  1.119830  4.810044  3.551002 -3.621536   
17206  0.0008  1.039211  0.268788  0.282907 -0.631976  5.732918 -2.032809   
22095  0.0107  1.673221  1.683247  1.490732 -0.799143  5.949979  1.238796   
3386  -0.0024  1.276583  3.618997  0.887167  5.938657  7.292157 -0.719356   
12353  0.0008  0.725808  3.079420  0.550588  2.251686  2.453580 -0.575291   
3432  -0.0068 -1.291289  5.117308  1.233501  0.403832  0.704562 -0.945536   
28592  0.0020  1.609066  5.217666  0.980613  1.761906  0.493248 -2.241785   
1828   0.0005 -0.012925  2.085193  0.799872 -2.055522  0.565912 -0.052163   

             x17        x18  

### x19

In [22]:
# replace dollar sign
dfCombine["x19"] = dfCombine["x19"].replace(to_replace="[\$]",value="",regex=True)
dfCombine.head(10)

y        x1         x2         x3        x4        x5        x6      x7  \
0  0.0  0.165254  18.060003  Wednesday  1.077380 -1.339233 -1.584341  0.0062   
1  1.0  2.441471  18.416307     Friday  1.482586  0.920817 -0.759931  0.0064   
2  1.0  4.427278  19.188092   Thursday  0.145652  0.366093  0.709962 -0.0008   
3  0.0  3.925235  19.901257    Tuesday  1.763602 -0.251926 -0.827461 -0.0057   
4  0.0  2.868802  22.202473     Sunday  3.405119  0.083162  1.381504  0.0109   
5  0.0  1.926053  19.524376   Saturday  3.702574  2.300792 -0.331106  0.0079   
6  0.0  0.094162  18.894755   Thursday  2.027918  1.580408  1.190932  0.0174   
7  0.0  0.921300  19.160713   Saturday -1.274067 -0.867023 -2.036134 -0.0106   
8  0.0  0.589354  21.627989  Wednesday  1.672593  0.158446  2.073026  0.0032   
9  0.0  1.016857  21.440000    Tuesday -2.619848  0.637245  1.357868  0.0091   

         x8        x9       x10       x12       x13       x15       x17  \
0  0.220784  1.816481  1.171788  4.644568  4.814885 -0.587361  4.278444   
1  1.192441  3.513950  1.419900  1.459868  1.443983 -1.165869  8.487132   
2  0.952323  0.782974 -1.247022  1.098525  1.216059  0.211685  8.601193   
3 -0.520756  1.825586  2.223038 -1.390239  3.962961 -2.046856  6.611554   
4 -0.732739  2.151990 -0.275406  7.230125  3.877312 -1.201565  9.103015   
5  3.085443  0.472701  1.943171 -3.582885  3.522031 -2.799700  7.283226   
6 -0.331730  2.178622 -0.454172  1.762009  2.125462 -0.907832  8.631917   
7  2.830838  1.206701  1.315474 -2.483859  4.224671 -2.123287  6.667980   
8 -0.908026 -0.064304  0.861625 -2.124964  1.903264 -3.364331  5.437052   
9  1.471836  0.811922 -0.699837  2.893282  1.603635 -3.283478  7.988783   

         x18                x19        x20        x21         x22       x23  \
0  77.709700  -908.650758424405   7.328029  36.601967  126.384294  1.264713   
1  70.808566   -1864.9622875143   6.796408   3.789176   93.037021  0.346201   
2  75.922820  -543.187402955527   3.802395   7.407996   71.022413  0.070968   
3  74.966925  -182.626380634258   7.728963   8.136213  121.610846  0.700954   
4  77.977156   967.007090837503   2.272765  24.452102   56.492317 -1.102387   
5  87.734440   271.163475698959  15.770180  44.874132   45.729121  0.320526   
6  71.282134   749.965009681221  14.220882  12.727490   40.145649 -0.196627   
7  78.224602   32.7532387681275  10.688299   4.209199   55.802747  1.255861   
8  81.846939   1079.84051477389   8.314540   1.074290   71.459845  1.175947   
9  89.462640  -1668.77796649785   4.055610  -0.328695   24.882245 -0.102272   

        x25       x26          x27        x28       x29  x31       x32  \
0  3.834082  2.352406   905.491445  14.779467  0.752524   no -0.597288   
1  4.345028  1.651579   868.784447  25.914680  1.445294   no  1.127376   
2  4.332644 -0.375737  1077.233497  13.177479  4.174536   no  2.605279   
3  7.294990 -0.603983  1051.655489  17.006528  2.347355   no  1.071202   
4  6.313827  0.429187   949.904947  16.962710  0.510985  yes  2.283921   
5  2.245217  4.303686  1187.333248   8.139008  2.233107   no  0.488196   
6  4.784640  1.051441   902.058622  17.881615  1.580413   no  2.424133   
7  3.560273  1.048213   990.811632  16.443997  0.241411  yes  0.123102   
8  2.640861 -0.255633  1214.321482  13.353094 -0.060317   no  0.368641   
9  2.166865 -0.795430  1244.523235   9.513302  0.273440   no  0.007047   

         x34       x35        x36        x37        x40       x43       x46  \
0  46.667221  3.159002  69.444841  32.450076  18.129499  1.227266  0.234654   
1  47.928569 -1.013712  61.289132  35.521824  11.389081 -0.822055  1.104631   
2  56.674425  0.108486  69.798923  30.684074  14.372443  0.150728 -2.561083   
3  59.154933  1.319711  65.408246  34.401290  13.191173 -0.965711 -2.123048   
4  46.445617  0.022747  66.662910  31.135261  17.210430  1.012093  3.477451   
5  45.559596  0.359074  76.835865  37.602918  12.628248  1.452704 -1.573853   
6  53.786798  2.610824  68.051870  32.855520  12.243767  1.8

### Round to 6 decimal places

In [23]:
dfCombine["x19"] = dfCombine["x19"].astype(float).round(6)
dfCombine.head(10)

y        x1         x2         x3        x4        x5        x6      x7  \
0  0.0  0.165254  18.060003  Wednesday  1.077380 -1.339233 -1.584341  0.0062   
1  1.0  2.441471  18.416307     Friday  1.482586  0.920817 -0.759931  0.0064   
2  1.0  4.427278  19.188092   Thursday  0.145652  0.366093  0.709962 -0.0008   
3  0.0  3.925235  19.901257    Tuesday  1.763602 -0.251926 -0.827461 -0.0057   
4  0.0  2.868802  22.202473     Sunday  3.405119  0.083162  1.381504  0.0109   
5  0.0  1.926053  19.524376   Saturday  3.702574  2.300792 -0.331106  0.0079   
6  0.0  0.094162  18.894755   Thursday  2.027918  1.580408  1.190932  0.0174   
7  0.0  0.921300  19.160713   Saturday -1.274067 -0.867023 -2.036134 -0.0106   
8  0.0  0.589354  21.627989  Wednesday  1.672593  0.158446  2.073026  0.0032   
9  0.0  1.016857  21.440000    Tuesday -2.619848  0.637245  1.357868  0.0091   

         x8        x9       x10       x12       x13       x15       x17  \
0  0.220784  1.816481  1.171788  4.644568  4.814885 -0.587361  4.278444   
1  1.192441  3.513950  1.419900  1.459868  1.443983 -1.165869  8.487132   
2  0.952323  0.782974 -1.247022  1.098525  1.216059  0.211685  8.601193   
3 -0.520756  1.825586  2.223038 -1.390239  3.962961 -2.046856  6.611554   
4 -0.732739  2.151990 -0.275406  7.230125  3.877312 -1.201565  9.103015   
5  3.085443  0.472701  1.943171 -3.582885  3.522031 -2.799700  7.283226   
6 -0.331730  2.178622 -0.454172  1.762009  2.125462 -0.907832  8.631917   
7  2.830838  1.206701  1.315474 -2.483859  4.224671 -2.123287  6.667980   
8 -0.908026 -0.064304  0.861625 -2.124964  1.903264 -3.364331  5.437052   
9  1.471836  0.811922 -0.699837  2.893282  1.603635 -3.283478  7.988783   

         x18          x19        x20        x21         x22       x23  \
0  77.709700  -908.650758   7.328029  36.601967  126.384294  1.264713   
1  70.808566 -1864.962288   6.796408   3.789176   93.037021  0.346201   
2  75.922820  -543.187403   3.802395   7.407996   71.022413  0.070968   
3  74.966925  -182.626381   7.728963   8.136213  121.610846  0.700954   
4  77.977156   967.007091   2.272765  24.452102   56.492317 -1.102387   
5  87.734440   271.163476  15.770180  44.874132   45.729121  0.320526   
6  71.282134   749.965010  14.220882  12.727490   40.145649 -0.196627   
7  78.224602    32.753239  10.688299   4.209199   55.802747  1.255861   
8  81.846939  1079.840515   8.314540   1.074290   71.459845  1.175947   
9  89.462640 -1668.777966   4.055610  -0.328695   24.882245 -0.102272   

        x25       x26          x27        x28       x29  x31       x32  \
0  3.834082  2.352406   905.491445  14.779467  0.752524   no -0.597288   
1  4.345028  1.651579   868.784447  25.914680  1.445294   no  1.127376   
2  4.332644 -0.375737  1077.233497  13.177479  4.174536   no  2.605279   
3  7.294990 -0.603983  1051.655489  17.006528  2.347355   no  1.071202   
4  6.313827  0.429187   949.904947  16.962710  0.510985  yes  2.283921   
5  2.245217  4.303686  1187.333248   8.139008  2.233107   no  0.488196   
6  4.784640  1.051441   902.058622  17.881615  1.580413   no  2.424133   
7  3.560273  1.048213   990.811632  16.443997  0.241411  yes  0.123102   
8  2.640861 -0.255633  1214.321482  13.353094 -0.060317   no  0.368641   
9  2.166865 -0.795430  1244.523235   9.513302  0.273440   no  0.007047   

         x34       x35        x36        x37        x40       x43       x46  \
0  46.667221  3.159002  69.444841  32.450076  18.129499  1.227266  0.234654   
1  47.928569 -1.013712  61.289132  35.521824  11.389081 -0.822055  1.104631   
2  56.674425  0.108486  69.798923  30.684074  14.372443  0.150728 -2.561083   
3  59.154933  1.319711  65.408246  34.401290  13.191173 -0.965711 -2.123048   
4  46.445617  0.022747  66.662910  31.135261  17.210430  1.012093  3.477451   
5  45.559596  0.359074  76.835865  37.602918  12.628248  1.452704 -1.573853   
6  53.786798  2.610824  68.051870  32.855520  12.243767  1.863568  0.977259   
7  54.209775  2.215420  70.230786  31.665864  1

### x31

In [24]:
# map 0 to no and 1 to yes
dfCombine["x31"] = dfCombine["x31"].map({"no":0,"yes":1})
dfCombine["x31"].unique()

array([0, 1], dtype=int64)

### x93

In [25]:
dfCombine["x93"] = dfCombine["x93"].map({"no":0,"yes":1})
dfCombine["x93"].unique()

array([0, 1], dtype=int64)

### Encode categorical variables

### x3

In [26]:
# days of the week
x3 = pd.get_dummies(dfCombine['x3'],prefix='cat_')
dfCombine = pd.merge(
    left=dfCombine,
    right=x3,
    left_index=True,
    right_index=True,
)

dfCombine.head(15)

y        x1         x2         x3        x4        x5        x6      x7  \
0   0.0  0.165254  18.060003  Wednesday  1.077380 -1.339233 -1.584341  0.0062   
1   1.0  2.441471  18.416307     Friday  1.482586  0.920817 -0.759931  0.0064   
2   1.0  4.427278  19.188092   Thursday  0.145652  0.366093  0.709962 -0.0008   
3   0.0  3.925235  19.901257    Tuesday  1.763602 -0.251926 -0.827461 -0.0057   
4   0.0  2.868802  22.202473     Sunday  3.405119  0.083162  1.381504  0.0109   
5   0.0  1.926053  19.524376   Saturday  3.702574  2.300792 -0.331106  0.0079   
6   0.0  0.094162  18.894755   Thursday  2.027918  1.580408  1.190932  0.0174   
7   0.0  0.921300  19.160713   Saturday -1.274067 -0.867023 -2.036134 -0.0106   
8   0.0  0.589354  21.627989  Wednesday  1.672593  0.158446  2.073026  0.0032   
9   0.0  1.016857  21.440000    Tuesday -2.619848  0.637245  1.357868  0.0091   
10  1.0  0.853152  20.300924  Wednesday -0.930155 -0.607742 -0.289608 -0.0052   
11  0.0  4.304882  19.302074     Sunday  2.470847 -1.852730  0.400993 -0.0025   
12  0.0  1.799370  20.725235     Friday -1.381245  0.716132  0.503553 -0.0045   
13  0.0  3.671553  22.207880    Tuesday  0.698724 -1.862047  2.153832  0.0068   
14  0.0  3.567967  16.644123  Wednesday -0.951598 -1.404246 -2.013598 -0.0137   

          x8        x9       x10       x12       x13       x15        x17  \
0   0.220784  1.816481  1.171788  4.644568  4.814885 -0.587361   4.278444   
1   1.192441  3.513950  1.419900  1.459868  1.443983 -1.165869   8.487132   
2   0.952323  0.782974 -1.247022  1.098525  1.216059  0.211685   8.601193   
3  -0.520756  1.825586  2.223038 -1.390239  3.962961 -2.046856   6.611554   
4  -0.732739  2.151990 -0.275406  7.230125  3.877312 -1.201565   9.103015   
5   3.085443  0.472701  1.943171 -3.582885  3.522031 -2.799700   7.283226   
6  -0.331730  2.178622 -0.454172  1.762009  2.125462 -0.907832   8.631917   
7   2.830838  1.206701  1.315474 -2.483859  4.224671 -2.123287   6.667980   
8  -0.908026 -0.064304  0.861625 -2.124964  1.903264 -3.364331   5.437052   
9   1.471836  0.811922 -0.699837  2.893282  1.603635 -3.283478   7.988783   
10 -0.040689  1.858641  1.491314 -1.600497  3.219078 -0.764358   8.769019   
11  0.907410  2.052139  1.186982  3.919136  4.463684 -2.649127   8.539390   
12  0.410077 -0.186681 -1.054522  1.021625  6.562940  0.201709  10.347098   
13  0.205453  1.309530  0.683070  1.433037  2.625811 -0.824623   6.134170   
14  1.578542  1.905247 -1.224807  1.194875  1.661333 -2.137908   3.995859   

          x18          x19        x20        x21         x22       x23  \
0   77.709700  -908.650758   7.328029  36.601967  126.384294  1.264713   
1   70.808566 -1864.962288   6.796408   3.789176   93.037021  0.346201   
2   75.922820  -543.187403   3.802395   7.407996   71.022413  0.070968   
3   74.966925  -182.626381   7.728963   8.136213  121.610846  0.700954   
4   77.977156   967.007091   2.272765  24.452102   56.492317 -1.102387   
5   87.734440   271.163476  15.770180  44.874132   45.729121  0.320526   
6   71.282134   749.965010  14.220882  12.727490   40.145649 -0.196627   
7   78.224602    32.753239  10.688299   4.209199   55.802747  1.255861   
8   81.846939  1079.840515   8.314540   1.074290   71.459845  1.175947   
9   89.462640 -1668.777966   4.055610  -0.328695   24.882245 -0.102272   
10  82.870050  1391.302605   7.099256   6.247910   59.310311 -1.107675   
11  78.858807 -1822.863099   7.000926  19.971813   95.353723 -1.060869   
12  69.524087  1038.634263   6.341097   0.817180   97.972638  1.783286   
13  72.430917  -377.495529   7.074143   7.676579   72.703358  0.520186   
14  89.192758   581.353320  10.061368   6.925499   71.816502 -0.724962   

         x25       x26          x27        x28       x29  x31       x32  \
0   3.834082  2.352406   905.491445  14.779467  0.752524    0 -0.597288   
1   4.345028  1.651579   868.784447  25.914680  1.445294    0  1.127376   
2   4.332644 -0.375737  1077.233497  13.177479  4.174536    0  2.6

#### x60

In [27]:
# months
x60 = pd.get_dummies(dfCombine['x60'],prefix='cat_')
dfCombine = pd.merge(
    left=dfCombine,
    right=x60,
    left_index=True,
    right_index=True,
)

dfCombine.head(15)

y        x1         x2         x3        x4        x5        x6      x7  \
0   0.0  0.165254  18.060003  Wednesday  1.077380 -1.339233 -1.584341  0.0062   
1   1.0  2.441471  18.416307     Friday  1.482586  0.920817 -0.759931  0.0064   
2   1.0  4.427278  19.188092   Thursday  0.145652  0.366093  0.709962 -0.0008   
3   0.0  3.925235  19.901257    Tuesday  1.763602 -0.251926 -0.827461 -0.0057   
4   0.0  2.868802  22.202473     Sunday  3.405119  0.083162  1.381504  0.0109   
5   0.0  1.926053  19.524376   Saturday  3.702574  2.300792 -0.331106  0.0079   
6   0.0  0.094162  18.894755   Thursday  2.027918  1.580408  1.190932  0.0174   
7   0.0  0.921300  19.160713   Saturday -1.274067 -0.867023 -2.036134 -0.0106   
8   0.0  0.589354  21.627989  Wednesday  1.672593  0.158446  2.073026  0.0032   
9   0.0  1.016857  21.440000    Tuesday -2.619848  0.637245  1.357868  0.0091   
10  1.0  0.853152  20.300924  Wednesday -0.930155 -0.607742 -0.289608 -0.0052   
11  0.0  4.304882  19.302074     Sunday  2.470847 -1.852730  0.400993 -0.0025   
12  0.0  1.799370  20.725235     Friday -1.381245  0.716132  0.503553 -0.0045   
13  0.0  3.671553  22.207880    Tuesday  0.698724 -1.862047  2.153832  0.0068   
14  0.0  3.567967  16.644123  Wednesday -0.951598 -1.404246 -2.013598 -0.0137   

          x8        x9       x10       x12       x13       x15        x17  \
0   0.220784  1.816481  1.171788  4.644568  4.814885 -0.587361   4.278444   
1   1.192441  3.513950  1.419900  1.459868  1.443983 -1.165869   8.487132   
2   0.952323  0.782974 -1.247022  1.098525  1.216059  0.211685   8.601193   
3  -0.520756  1.825586  2.223038 -1.390239  3.962961 -2.046856   6.611554   
4  -0.732739  2.151990 -0.275406  7.230125  3.877312 -1.201565   9.103015   
5   3.085443  0.472701  1.943171 -3.582885  3.522031 -2.799700   7.283226   
6  -0.331730  2.178622 -0.454172  1.762009  2.125462 -0.907832   8.631917   
7   2.830838  1.206701  1.315474 -2.483859  4.224671 -2.123287   6.667980   
8  -0.908026 -0.064304  0.861625 -2.124964  1.903264 -3.364331   5.437052   
9   1.471836  0.811922 -0.699837  2.893282  1.603635 -3.283478   7.988783   
10 -0.040689  1.858641  1.491314 -1.600497  3.219078 -0.764358   8.769019   
11  0.907410  2.052139  1.186982  3.919136  4.463684 -2.649127   8.539390   
12  0.410077 -0.186681 -1.054522  1.021625  6.562940  0.201709  10.347098   
13  0.205453  1.309530  0.683070  1.433037  2.625811 -0.824623   6.134170   
14  1.578542  1.905247 -1.224807  1.194875  1.661333 -2.137908   3.995859   

          x18          x19        x20        x21         x22       x23  \
0   77.709700  -908.650758   7.328029  36.601967  126.384294  1.264713   
1   70.808566 -1864.962288   6.796408   3.789176   93.037021  0.346201   
2   75.922820  -543.187403   3.802395   7.407996   71.022413  0.070968   
3   74.966925  -182.626381   7.728963   8.136213  121.610846  0.700954   
4   77.977156   967.007091   2.272765  24.452102   56.492317 -1.102387   
5   87.734440   271.163476  15.770180  44.874132   45.729121  0.320526   
6   71.282134   749.965010  14.220882  12.727490   40.145649 -0.196627   
7   78.224602    32.753239  10.688299   4.209199   55.802747  1.255861   
8   81.846939  1079.840515   8.314540   1.074290   71.459845  1.175947   
9   89.462640 -1668.777966   4.055610  -0.328695   24.882245 -0.102272   
10  82.870050  1391.302605   7.099256   6.247910   59.310311 -1.107675   
11  78.858807 -1822.863099   7.000926  19.971813   95.353723 -1.060869   
12  69.524087  1038.634263   6.341097   0.817180   97.972638  1.783286   
13  72.430917  -377.495529   7.074143   7.676579   72.703358  0.520186   
14  89.192758   581.353320  10.061368   6.925499   71.816502 -0.724962   

         x25       x26          x27        x28       x29  x31       x32  \
0   3.834082  2.352406   905.491445  14.779467  0.752524    0 -0.597288   
1   4.345028  1.651579   868.784447  25.914680  1.445294    0  1.127376   
2   4.332644 -0.375737  1077.233497  13.177479  4.174536    0  2.6

#### x65

In [28]:
# insurance name
x65 = pd.get_dummies(dfCombine['x65'],prefix='cat_')
dfCombine = pd.merge(
    left=dfCombine,
    right=x65,
    left_index=True,
    right_index=True,
)

dfCombine.head(15)

y        x1         x2         x3        x4        x5        x6      x7  \
0   0.0  0.165254  18.060003  Wednesday  1.077380 -1.339233 -1.584341  0.0062   
1   1.0  2.441471  18.416307     Friday  1.482586  0.920817 -0.759931  0.0064   
2   1.0  4.427278  19.188092   Thursday  0.145652  0.366093  0.709962 -0.0008   
3   0.0  3.925235  19.901257    Tuesday  1.763602 -0.251926 -0.827461 -0.0057   
4   0.0  2.868802  22.202473     Sunday  3.405119  0.083162  1.381504  0.0109   
5   0.0  1.926053  19.524376   Saturday  3.702574  2.300792 -0.331106  0.0079   
6   0.0  0.094162  18.894755   Thursday  2.027918  1.580408  1.190932  0.0174   
7   0.0  0.921300  19.160713   Saturday -1.274067 -0.867023 -2.036134 -0.0106   
8   0.0  0.589354  21.627989  Wednesday  1.672593  0.158446  2.073026  0.0032   
9   0.0  1.016857  21.440000    Tuesday -2.619848  0.637245  1.357868  0.0091   
10  1.0  0.853152  20.300924  Wednesday -0.930155 -0.607742 -0.289608 -0.0052   
11  0.0  4.304882  19.302074     Sunday  2.470847 -1.852730  0.400993 -0.0025   
12  0.0  1.799370  20.725235     Friday -1.381245  0.716132  0.503553 -0.0045   
13  0.0  3.671553  22.207880    Tuesday  0.698724 -1.862047  2.153832  0.0068   
14  0.0  3.567967  16.644123  Wednesday -0.951598 -1.404246 -2.013598 -0.0137   

          x8        x9       x10       x12       x13       x15        x17  \
0   0.220784  1.816481  1.171788  4.644568  4.814885 -0.587361   4.278444   
1   1.192441  3.513950  1.419900  1.459868  1.443983 -1.165869   8.487132   
2   0.952323  0.782974 -1.247022  1.098525  1.216059  0.211685   8.601193   
3  -0.520756  1.825586  2.223038 -1.390239  3.962961 -2.046856   6.611554   
4  -0.732739  2.151990 -0.275406  7.230125  3.877312 -1.201565   9.103015   
5   3.085443  0.472701  1.943171 -3.582885  3.522031 -2.799700   7.283226   
6  -0.331730  2.178622 -0.454172  1.762009  2.125462 -0.907832   8.631917   
7   2.830838  1.206701  1.315474 -2.483859  4.224671 -2.123287   6.667980   
8  -0.908026 -0.064304  0.861625 -2.124964  1.903264 -3.364331   5.437052   
9   1.471836  0.811922 -0.699837  2.893282  1.603635 -3.283478   7.988783   
10 -0.040689  1.858641  1.491314 -1.600497  3.219078 -0.764358   8.769019   
11  0.907410  2.052139  1.186982  3.919136  4.463684 -2.649127   8.539390   
12  0.410077 -0.186681 -1.054522  1.021625  6.562940  0.201709  10.347098   
13  0.205453  1.309530  0.683070  1.433037  2.625811 -0.824623   6.134170   
14  1.578542  1.905247 -1.224807  1.194875  1.661333 -2.137908   3.995859   

          x18          x19        x20        x21         x22       x23  \
0   77.709700  -908.650758   7.328029  36.601967  126.384294  1.264713   
1   70.808566 -1864.962288   6.796408   3.789176   93.037021  0.346201   
2   75.922820  -543.187403   3.802395   7.407996   71.022413  0.070968   
3   74.966925  -182.626381   7.728963   8.136213  121.610846  0.700954   
4   77.977156   967.007091   2.272765  24.452102   56.492317 -1.102387   
5   87.734440   271.163476  15.770180  44.874132   45.729121  0.320526   
6   71.282134   749.965010  14.220882  12.727490   40.145649 -0.196627   
7   78.224602    32.753239  10.688299   4.209199   55.802747  1.255861   
8   81.846939  1079.840515   8.314540   1.074290   71.459845  1.175947   
9   89.462640 -1668.777966   4.055610  -0.328695   24.882245 -0.102272   
10  82.870050  1391.302605   7.099256   6.247910   59.310311 -1.107675   
11  78.858807 -1822.863099   7.000926  19.971813   95.353723 -1.060869   
12  69.524087  1038.634263   6.341097   0.817180   97.972638  1.783286   
13  72.430917  -377.495529   7.074143   7.676579   72.703358  0.520186   
14  89.192758   581.353320  10.061368   6.925499   71.816502 -0.724962   

         x25       x26          x27        x28       x29  x31       x32  \
0   3.834082  2.352406   905.491445  14.779467  0.752524    0 -0.597288   
1   4.345028  1.651579   868.784447  25.914680  1.445294    0  1.127376   
2   4.332644 -0.375737  1077.233497  13.177479  4.174536    0  2.6

In [29]:
dfTrainClean = dfCombine.iloc[:40000,]
dfTestClean = dfCombine.iloc[40000:,]

### Logistic regression (classification) & neural network

class sklearn.linear_model.LogisticRegression(penalty='l2',dual=False,tol=0.0001,C=1.0,fit_intercept=True,
intercept_scaling=1,class_weight=None,random_state=None,solver='lbfgs',max_iter=100,multi_class='auto', 
verbose=0,warm_start=False,n_jobs=None,l1_ratio=None)

class sklearn.neural_network.MLPClassifier(hidden_layer_sizes=(100, ), activation='relu', *, 
solver='adam',alpha=0.0001,batch_size='auto', learning_rate='constant', learning_rate_init=0.001,
power_t=0.5,max_iter=200,shuffle=True, random_state=None, tol=0.0001, verbose=False, 
warm_start=False, momentum=0.9,nesterovs_momentum=True, early_stopping=False, 
validation_fraction=0.1,beta_1=0.9,beta_2=0.999,epsilon=1e-08,n_iter_no_change=10, max_fun=15000)

In [31]:
dfCombine.columns

Index(['y', 'x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7', 'x8', 'x9', 'x10', 'x12',
       'x13', 'x15', 'x17', 'x18', 'x19', 'x20', 'x21', 'x22', 'x23', 'x25',
       'x26', 'x27', 'x28', 'x29', 'x31', 'x32', 'x34', 'x35', 'x36', 'x37',
       'x40', 'x43', 'x46', 'x47', 'x48', 'x50', 'x51', 'x53', 'x56', 'x58',
       'x59', 'x60', 'x62', 'x63', 'x65', 'x66', 'x67', 'x68', 'x69', 'x70',
       'x71', 'x72', 'x73', 'x81', 'x82', 'x84', 'x87', 'x90', 'x92', 'x93',
       'x94', 'x97', 'x98', 'x100', 'cat__Friday', 'cat__Monday',
       'cat__Saturday', 'cat__Sunday', 'cat__Thursday', 'cat__Tuesday',
       'cat__Wednesday', 'cat__April', 'cat__August', 'cat__December',
       'cat__February', 'cat__January', 'cat__July', 'cat__June', 'cat__March',
       'cat__May', 'cat__November', 'cat__October', 'cat__September',
       'cat__allstate', 'cat__esurance', 'cat__farmers', 'cat__geico',
       'cat__progressive'],
      dtype='object')

In [32]:
features = [
    'x1','x2','x4','x5','x6','x7','x8','x9','x10','x12',
    'x13','x15','x17','x18', 'x19', 'x20', 'x21','x22', 'x23', 'x25',
    'x26','x27','x28','x29', 'x31', 'x32', 'x34','x35', 'x36', 'x37',
    'x40','x43','x46','x47', 'x48', 'x50', 'x51','x53', 'x56', 'x58',
    'x59','x62','x63','x66', 'x67', 'x68', 'x69','x70',
    'x71','x72','x73','x81', 'x82', 'x84', 'x87','x90','x92','x93',
    'x94','x97','x98', 'x100', 'cat__Friday', 'cat__Monday',
    'cat__Saturday','cat__Sunday','cat__Thursday', 'cat__Tuesday',
    'cat__Wednesday','cat__April','cat__August','cat__December',
    'cat__February','cat__January','cat__July','cat__June', 'cat__March',
    'cat__May','cat__November', 'cat__October', 'cat__September',
    'cat__allstate','cat__esurance', 'cat__farmers', 'cat__geico',
    'cat__progressive']

# x4 x6 x9 x10 x12 x17 x22 x23 x34 x48 x53 x59 x69 x73 x81 x92 x97
dropFeatures = [
    'x1','x2','x5','x7','x8',
    'x13','x15','x18','x19', 'x20','x21','x25',
    'x26','x27','x28','x29', 'x31','x32','x35', 'x36', 'x37',
    'x40','x43','x46','x47', 'x50','x51', 'x56','x58',
    'x62','x63','x66', 'x67', 'x68','x70',
    'x71','x72','x82','x84', 'x87','x90','x93',
    'x94','x98', 'x100', 'cat__Friday', 'cat__Monday',
    'cat__Saturday','cat__Sunday','cat__Thursday', 'cat__Tuesday',
    'cat__Wednesday','cat__April','cat__August','cat__December',
    'cat__February','cat__January','cat__July','cat__June', 'cat__March',
    'cat__May','cat__November', 'cat__October', 'cat__September',
    'cat__allstate','cat__esurance', 'cat__farmers', 'cat__geico',
    'cat__progressive']

### Scale data

In [35]:
scalerX = StandardScaler()
scalerX.fit(dfTrainClean[dropFeatures])
xTrainScaled = scalerX.transform(dfTrainClean[dropFeatures])
xTestScaled = scalerX.transform(dfTestClean[dropFeatures])

In [36]:
# logistic regression
logReg = LogisticRegression(C=7.7777,solver="lbfgs",max_iter=1111,tol=0.000003333, \
    multi_class="ovr",class_weight=None)

logReg.fit(xTrainScaled.astype(float),dfTrainClean["y"].astype(int))

LogisticRegression(C=7.7777, max_iter=1111, multi_class='ovr', tol=3.333e-06)

In [37]:
# neural networks
mlp = MLPClassifier(hidden_layer_sizes=(76,76,),random_state=None,alpha=0.1111, \
    solver='sgd',learning_rate="adaptive",activation="relu",n_iter_no_change=10, \
    tol=0.00001111,learning_rate_init=0.001111,max_iter=253,momentum=0.9111, \
    batch_size="auto",shuffle=True,warm_start=False,early_stopping=False,validation_fraction=0.1)

mlp.fit(xTrainScaled.astype(float),dfTrainClean["y"].astype(int))

C:\Users\TamLe\anaconda3\envs\sports\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:614: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (253) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPClassifier(alpha=0.1111, hidden_layer_sizes=(76, 76),
              learning_rate='adaptive', learning_rate_init=0.001111,
              max_iter=253, momentum=0.9111, solver='sgd', tol=1.111e-05)

### Predictions

In [38]:
yPredTrainLog = logReg.predict(xTrainScaled.astype(float))
yPredTestLog = logReg.predict(xTestScaled.astype(float))
yProbTrainLog = logReg.predict_proba(xTrainScaled.astype(float))
yProbTestLog = logReg.predict_proba(xTestScaled.astype(float))

In [39]:
yPredTrainMLP = mlp.predict(xTrainScaled.astype(float))
yPredTestMLP = mlp.predict(xTestScaled.astype(float))
yProbTrainMLP = mlp.predict_proba(xTrainScaled.astype(float))
yProbTestMLP = mlp.predict_proba(xTestScaled.astype(float))

In [40]:
pd.DataFrame(yPredTrainLog).value_counts()

0    39238
1      762
dtype: int64

In [41]:
pd.DataFrame(dfTrain["y"]).value_counts()

y
0    34197
1     5803
dtype: int64

### Probabilities on training set

In [42]:
pd.DataFrame(yProbTrainLog).head(20)

0         1
0   0.959971  0.040029
1   0.658801  0.341199
2   0.828818  0.171182
3   0.549976  0.450024
4   0.983968  0.016032
5   0.888512  0.111488
6   0.958273  0.041727
7   0.859887  0.140113
8   0.924804  0.075196
9   0.854813  0.145187
10  0.703103  0.296897
11  0.647305  0.352695
12  0.953554  0.046446
13  0.959665  0.040335
14  0.962788  0.037212
15  0.643323  0.356677
16  0.904401  0.095599
17  0.789453  0.210547
18  0.980343  0.019657
19  0.698038  0.301962

In [43]:
pd.DataFrame(yProbTestLog).tail(25)

0         1
9975  0.946557  0.053443
9976  0.909108  0.090892
9977  0.808916  0.191084
9978  0.656960  0.343040
9979  0.330718  0.669282
9980  0.838664  0.161336
9981  0.948758  0.051242
9982  0.948408  0.051592
9983  0.943232  0.056768
9984  0.768390  0.231610
9985  0.799879  0.200121
9986  0.727411  0.272589
9987  0.768810  0.231190
9988  0.941240  0.058760
9989  0.763365  0.236635
9990  0.957531  0.042469
9991  0.951890  0.048110
9992  0.910506  0.089494
9993  0.908765  0.091235
9994  0.932012  0.067988
9995  0.971479  0.028521
9996  0.731335  0.268665
9997  0.898240  0.101760
9998  0.780030  0.219970
9999  0.971902  0.028098

In [44]:
pd.DataFrame(yProbTestLog).iloc[:,[1]].head(15)

1
0   0.020053
1   0.027552
2   0.259752
3   0.177349
4   0.013140
5   0.057507
6   0.274311
7   0.059734
8   0.326363
9   0.623363
10  0.224240
11  0.406556
12  0.677551
13  0.187008
14  0.188720

### Save results

In [45]:
pd.DataFrame(yProbTestLog).iloc[:,[1]].to_csv("glmresults.csv",header=False,index=False)
pd.DataFrame(yProbTestMLP).iloc[:,[1]].to_csv("nonglmresults.csv",header=False,index=False)

In [46]:
print(len(yProbTestMLP))

10000


### Statsmodels

In [49]:
xInclConstant = sm.add_constant(dfTrainClean[dropFeatures])
model = sm.Logit(dfTrainClean["y"].astype(int),xInclConstant.astype(float))
# logistic regression
results = model.fit()
results.summary()

Optimization terminated successfully.
         Current function value: 0.359918
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                      y   No. Observations:                40000
Model:                          Logit   Df Residuals:                    39933
Method:                           MLE   Df Model:                           66
Date:                Fri, 04 Jun 2021   Pseudo R-squ.:                  0.1308
Time:                        18:55:35   Log-Likelihood:                -14397.
converged:                       True   LL-Null:                       -16563.
Covariance Type:            nonrobust   LLR p-value:                     0.000
====================================================================================
                       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------
const                3.5752   2.06e+06   1.73e-06      1.000   -4.04e+06    4.04e+06
x1                   0.0259      0.008      3.294      0.001       0.011       0.041
x2                   0.0954      0.022      4.407      0.000       0.053       0.138
x5                   0.0446      0.016      2.826      0.005       0.014       0.076
x7                 -74.1062      1.636    -45.290      0.000     -77.313     -70.899
x8                   0.1096      0.018      6.179      0.000       0.075       0.144
x13                  0.0194      0.009      2.108      0.035       0.001       0.037
x15                 -0.0205      0.014     -1.494      0.135      -0.047       0.006
x18                 -0.0094      0.005     -2.034      0.042      -0.018      -0.000
x19                 -0.0001   1.51e-05     -9.304      0.000      -0.000      -0.000
x20                 -0.0301      0.006     -4.903      0.000      -0.042      -0.018
x21                  0.0012      0.001      1.199      0.231      -0.001       0.003
x25                  0.0058      0.006      0.910      0.363      -0.007       0.018
x26                  0.0143      0.014      1.050      0.294      -0.012       0.041
x27                 -0.0004      0.000     -1.429      0.153      -0.001       0.000
x28                 -0.0973      0.012     -7.867      0.000      -0.122      -0.073
x29                 -0.0202      0.013     -1.574      0.115      -0.045       0.005
x31                 -0.5720      0.059     -9.740      0.000      -0.687      -0.457
x32                  0.0144      0.009      1.642      0.101      -0.003       0.032
x35                  0.0243      0.015      1.670      0.095      -0.004       0.053
x36                 -0.0115      0.010     -1.209      0.227      -0.030       0.007
x37                  0.0623      0.009      6.603      0.000       0.044       0.081
x40                 -0.1068      0.008    -13.138      0.000      -0.123      -0.091
x43                 -0.0168      0.015     -1.118      0.264      -0.046       0.013
x46                 -0.0549      0.012     -4.517      0.000      -0.079      -0.031
x47                 -0.1394      0.008    -16.839      0.000      -0.156      -0.123
x50              -9.406e-05   2.48e-05     -3.796      0.000      -0.000   -4.55e-05
x51                  0.0436      0.010      4.290      0.000       0.024       0.063
x56                 -0.0230      0.011     -2.158      0.031      -0.044      -0.002
x58                  0.0007      0.000      1.596      0.110      -0.000       0.002
x62                 -0.0533      0.013     -3.975      0.000      -0.080      -0.027
x63                  0.0444      0.013      3.485      0.000       0.019       0.069
x66                  0.0689      0.016      4.193      0.000       0.037       0.101
x67                 -0.0254      0.014     -1.867      0.062      -0.052       0.001
x68                  0.0145      0.003      4.265      0.000       0.008       0.021
x70        

### AUC curve on training set
There is no dependent variable in test set.

### Logistic regresion score

In [53]:
yTrainClean = dfTrainClean["y"].to_numpy()
print(yTrainClean)
falsePositiveRate,truePositiveRate,threshold = metrics.roc_curve(yTrainClean,yProbTrainLog[:,1])

rocAucScore = metrics.roc_auc_score(yTrainClean,yProbTrainLog[:,1])
print("auc score: %s" % rocAucScore)
logLoss = log_loss(yTrainClean,yProbTrainLog[:,1])
print("log loss: %s" % logLoss)

[0. 1. 1. ... 0. 0. 0.]
auc score: 0.7527958790394674
log loss: 0.35991762698348395


### Neural networks score

In [54]:
yTrainClean = dfTrainClean["y"].to_numpy()
print(yTrainClean)
falsePositiveRate,truePositiveRate,threshold = metrics.roc_curve(yTrainClean,yProbTrainMLP[:,1])

rocAucScore = metrics.roc_auc_score(yTrainClean,yProbTrainMLP[:,1])
print("auc score: %s" % rocAucScore)
logLoss = log_loss(yTrainClean,yProbTrainMLP[:,1])
print("log loss: %s" % logLoss)

[0. 1. 1. ... 0. 0. 0.]
auc score: 0.8323815264437423
log loss: 0.3154787944414712
